File: AnkiDroid/src/main/java/com/ichi2/anki/exception/DeckRenameException.java

package com.ichi2.anki.exception;

import android.content.res.Resources;

import com.ichi2.anki.R;

public class DeckRenameException extends Exception {

    public static final int ALREADY_EXISTS = 0;
    public static final int FILTERED_NOSUBDECKS = 1;

    private final int mErrorCode;

    public DeckRenameException(int errorCode) {
        super();
        mErrorCode = errorCode;
    }

    public String getLocalizedMessage(Resources res) {
        switch (mErrorCode) {
            case ALREADY_EXISTS:
                return res.getString(R.string.decks_rename_exists);
            case FILTERED_NOSUBDECKS:
                return res.getString(R.string.decks_rename_filtered_nosubdecks);
            default:
                return "";
        }
    }
}

File: AnkiDroid/src/main/java/com/ichi2/libanki/backend/exception/DeckRenameException.kt
//noinspection MissingCopyrightHeader #8659
package com.ichi2.libanki.backend.exception

class DeckRenameException(val errorCode: Int) : Exception() {
    // region only if FILTERED_NOSUBDECKS
    private var filteredAncestorName: String? = null
    private var deckName: String? = null
    // endregion

    override val message: String?
        get() = if (errorCode == FILTERED_NOSUBDECKS) {
            "Deck $deckName has filtered ancestor $filteredAncestorName"
        } else {
            super.message
        }

    companion object {
        const val ALREADY_EXISTS = 0
        const val FILTERED_NOSUBDECKS = 1

        /** Generates a [DeckRenameException] with additional information in the message */
        fun filteredAncestor(
            deckName: String?,
            filteredAncestorName: String?,
        ): DeckRenameException = DeckRenameException(FILTERED_NOSUBDECKS).apply {
            this.filteredAncestorName = filteredAncestorName
            this.deckName = deckName
        }
    }
}
